scipy/conftest.py

```diff
@@ -160,14 +184,18 @@ def wrapper(func):
         )
         # method gets there as a function so we cannot use inspect.ismethod
         if '.' in func.__qualname__:
-            def wrapped(self, *args, xp, **kwargs):
+            @wraps(func)
+            def wrapped(self, *args, **kwargs):
+                xp = kwargs["xp"]
                 if xp.__name__ == backend:
                     pytest.skip(reason=reason)
-                return func(self, *args, xp, **kwargs)
+                return func(self, *args, **kwargs)
         else:
-            def wrapped(*args, xp, **kwargs):  # type: ignore[misc]
+            @wraps(func)
+            def wrapped(*args, **kwargs):
+                xp = kwargs["xp"]
                 if xp.__name__ == backend:
                     pytest.skip(reason=reason)
-                return func(*args, xp, **kwargs)
+                return func(*args, **kwargs)
```

Comment:
If I do this to test on this branch:

```diff
--- a/scipy/cluster/tests/test_vq.py
+++ b/scipy/cluster/tests/test_vq.py
@@ -282,7 +282,8 @@ class TestKMean:
 
     @array_api_compatible
     @skip_if_array_api_backend('numpy.array_api')
-    def test_kmeans2_rank1_2(self, xp):
+    @pytest.mark.parametrize("junk", [1, 2])
+    def test_kmeans2_rank1_2(self, xp, junk):
         data = xp.asarray(TESTDATA_2D)
         data1 = data[:, 0]
         kmeans2(data1, xp.asarray(2), iter=1)
```

`python dev.py test -j 32 -b all -- -k "test_kmeans2_rank1_2"`

I get: `no tests ran in 18.76s`, which doesn't seem quite right--shouldn't we still run for `torch` here? This is particularly complex, because it looks like `class TestKMean` also has a class-level GPU skip. It also seems to me like we should report the skip(s) with reasons(s) rather than "no tests ran."

If I delete the class-level skip for `TestKMean` in this branch and `main`, the behavior in this branch is indeed superior than a cherry-pick of the same diff onto `main`, where the original problem of breaking parametrize returns:

```
TypeError: TestKMean.test_kmeans2_rank1_2() missing 1 required positional argument: 'junk'
```

So, this is a bit long-winded, but I think it may be important to consider three things here:
1) This is probably a solid improvement (I'm slightly biased, obviously)
2) It isn't the full story yet, and we should probably open an issue for the behavior of class level + function level custom decorators we add, though I'd also be included to avoid the class-level ones as much as possible...
3) This is getting pretty complex to review and test locally already--I hate to say it, but I wonder if we're going to need tests for our new test infrastructure for sanity reasons (this could maybe be discussed in the issue). Would still be messy because devices not available on CI. Obviously, would be nice if this infra could be maintained externally/community level, but that's perhaps farther out...

Reply:
> I get: no tests ran in 18.76s, which doesn't seem quite right--shouldn't we still run for torch here?

I suspect that this is down to my use of `if torch.cuda.is_available():`. I hoped that this would not skip pytorch-cpu but it seems it is. I don't know enough about `torch` to say how to skip only on GPU then...

Reply:
Hmm maybe not, I still get `no tests ran` with `-b numpy` rather than `-b all`. I'm still not confident in `if torch.cuda.is_available():` though.

Reply:
I think we can probably agree this is getting confusing though.

Reply:
The decorator of this branch in combination with the class-level skip is completely broken: `python dev.py test -b all -t scipy/cluster/tests/test_vq.py::TestKMean` gives
```
========================================================================================== no tests ran in 0.04s ===========================================================================================
ERROR: not found: /home/lucas/dev/myscipy/build-install/lib/python3.10/site-packages/scipy/cluster/tests/test_vq.py::TestKMean
(no name '/home/lucas/dev/myscipy/build-install/lib/python3.10/site-packages/scipy/cluster/tests/test_vq.py::TestKMean' in any of [<Module test_vq.py>])
```

Reply:
Yeah, I was a getting a bit worried that we'll end up going in circles because we'll fix N behaviors and cause just 1 other thing to break instead, so that's why the "testing the tests" idea is suggested, but ugh.

Reply:
for what it's worth, class-level skips with `skip_if_array_api_backend` are currently completely broken on `main` too, giving the same error as above.